### PR TITLE
New FHEW bootstrapping method LMKCDEY (ia.cr/2022/198)

### PR DESCRIPTION
**Pull request by dsuponitskiy**
This PR includes all changes made for the original LMKCDEY PR submitted by Yongwoo Lee (see https://github.com/openfheorg/openfhe-development/pull/336).
The following text was taken from that original PR:

The pull request implements the bootstrapping of an FHEW-like scheme, based on LMKCDEY (ia.cr/2022/198).

LMKCDEY requires two different types of keys: n RGSW(X^si) and w+1 (~10) Automorphism keys.
Due to the complexity of making a distinguished member of class, I have used std::make_shared(1, 2, n) for both keys ([0, 0, 0:n] for RGSW keys and [0, 0, 0:w+1] for Automorphism keys) which has resulted in a small inefficiency.

Additionally, round-to-odd has not been implemented, as it would require significant changes to the code structure. Instead, the implementation uses the nearest odd integer from the given input, which may slightly increase the noise.

Lastly, I have hardcoded the window parameter w as 10, which is efficient enough while maintaining a small key size.

I have thoroughly tested my changes and ensured that they do not introduce any new issues or conflicts.

I would greatly appreciate it if you could review my pull request at your earliest convenience.
Please let me know if there are any further changes or modifications that I can make to improve the code or address any issues.

Thank you for your time and consideration. I look forward to your response.

### CODE REVIEW

**Review by Yongwoo-Lee-ccl — APPROVED** [review]
Checked the PR. No error found.

**Review by kimandrik — COMMENTED** [review]
Looks good to me. Would be great to have a benchmark for the new method

Also would be good to have window as a parameter in binfhecontext

**Comment by Yongwoo-Lee-ccl** [comment]
> Looks good to me. Would be great to have a benchmark for the new method
> 
> Also would be good to have window as a parameter in binfhecontext

Thanks for the review. I'll add benchmark, and find a good way to add window parameter

**Comment by Yongwoo-Lee-ccl** [comment]
Now benchmark is available. 
In my M1 mac, GINX with STD128_NAND took 96067 us and LMK+ with STD128_LMKCDEY_NAND took 65138 us.

File: src/binfhe/lib/rgsw-acc-lmkcdey.cpp [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+
+        (*ek)[0][0][i] = KeyGenLMKCDEY(params, skNTT, s);
+    }
+    
+    NativeInteger gen = NativeInteger(5);
+    
+    (*ek)[0][1][0] = KeyGenAuto(params, skNTT, 2*N - gen.ConvertToInt());
+
+    // m_window: window size, consider parameterization in the future
+    for (size_t i = 1; i < m_window+1; i++)
```
Comment (yspolyakov) [inline src/binfhe/lib/rgsw-acc-lmkcdey.cpp:67]: Why not use OMP parallelization here?

**Comment by yspolyakov** [comment]
Additional review comments:

1. All unit tests that currently exist for both `GINX` and `AP` should be added for the `LMKCDEY` mode. @dsuponitskiy please help as needed. Serialization should be tested as part of this.
2. Add  `STD128_LMKCDEY` to binfhe-constants-impl.cpp. 
3. Generation of m_logGen should be added to the Precompute method (rgsw-cryptoparameters.cpp). Otherwise we will have problems when deserializing the binfhe cryptocontext.

File: src/binfhe/lib/rgsw-acc-lmkcdey.cpp (hunk 1 above)
Comment (Yongwoo-Lee-ccl) [reply src/binfhe/lib/rgsw-acc-lmkcdey.cpp:67]: Of course we can use parallelization here. I'll add it.

**Comment by yspolyakov** [comment]
@Yongwoo-Lee-ccl Is the PR ready for re-review? Which items have been addressed? @dsuponitskiy-duality can add unit tests.

> 1. All unit tests that currently exist for both GINX and AP should be added for the LMKCDEY mode. @dsuponitskiy please help as needed. Serialization should be tested as part of this.
> 2. Add STD128_LMKCDEY to binfhe-constants-impl.cpp.
> 3. Generation of m_logGen should be added to the Precompute method (rgsw-cryptoparameters.cpp). Otherwise we will have problems when deserializing the binfhe cryptocontext.
> 4. Revise the parameters for STD128_LMKCDEY.

**Comment by Yongwoo-Lee-ccl** [comment]
@yspolyakov @dsuponitskiy-duality Sorry for the late answer. I addressed all the items mentioned above. 

1. Please find unit tests in benchmark/src/binfhe-lmkcdey.cpp
2. STD128_LMKCDEY is now added to binfhe-constants-impl.cpp.
3. Generation of m_logGen is done in the Precompute method
4. Gaussian secret support is added (similar to pke.)

File: src/binfhe/lib/binfhe-constants-impl.cpp [hunk 2]
```diff
@@ -118,11 +121,48 @@ std::ostream& operator<<(std::ostream& s, BINFHE_METHOD f) {
         case GINX:
             s << "CGGI";
             break;
+        case LMKCDEY:
+            s << "LMKCDEY";
+            break;
+        default:
+            s << "UKNOWN";
+            break;
+    }
```
Comment (yspolyakov) [inline src/binfhe/lib/binfhe-constants-impl.cpp:130]: Correct the typo. Should be UKNOWN.

File: src/binfhe/lib/binfhe-constants-impl.cpp [hunk 3]
```diff
@@ -118,11 +121,48 @@ std::ostream& operator<<(std::ostream& s, BINFHE_METHOD f) {
...
+            s << "XNOR_FAST";
+            break;
+        case XOR:
+            s << "XOR";
+            break;
+        case XNOR:
+            s << "XNOR";
+            break;
         default:
             s << "UKNOWN";
```
Comment (yspolyakov) [inline src/binfhe/lib/binfhe-constants-impl.cpp:161]: Correct the typo.

File: src/binfhe/lib/binfhe-constants-impl.cpp [hunk 4]
```diff
@@ -118,11 +121,48 @@ std::ostream& operator<<(std::ostream& s, BINFHE_METHOD f) {
...
+            break;
+        case XNOR_FAST:
+            s << "XNOR_FAST";
+            break;
+        case XOR:
+            s << "XOR";
+            break;
+        case XNOR:
+            s << "XNOR";
+            break;
```
Comment (yspolyakov) [inline src/binfhe/lib/binfhe-constants-impl.cpp:159]: Change to UNKNOWN

**Review by yspolyakov — APPROVED** [review]

**Pull request merged**